### PR TITLE
ads-replacement: adding gRPC-mux support for ADS config replacement

### PR DESCRIPTION
**Pull request by adisuissa**
Commit Message: ads-replacement: adding gRPC-mux support for ADS config replacement
Additional Description:
Adds support for replacing the ADS config in runtime. This PR includes the changes for the gRPC-Mux (Sotw and Delta).
First part of #35956.

Risk Level: low - requires another PR that enables this
Testing: Added unit and integration tests.
Docs Changes: N/A
Release Notes: N/A - not as part of this PR.
Platform Specific Features: N/A

### CODE REVIEW

**Comment by adisuissa** [comment]
Assigning @nezdolik for a maintainer review.
/assign @nezdolik

File: source/extensions/config_subscription/grpc/grpc_mux_failover.h
Comment (nezdolik) [inline source/extensions/config_subscription/grpc/grpc_mux_failover.h:9]: is forward declaration really needed here?

File: source/extensions/config_subscription/grpc/grpc_mux_impl.cc
Comment (nezdolik) [inline source/extensions/config_subscription/grpc/grpc_mux_impl.cc:87]: nit: when move semantics is intended prefer to declare arg types as rvalue references, e.g. `Grpc::RawAsyncClientPtr&&` etc

File: source/extensions/config_subscription/grpc/grpc_mux_impl.cc
Comment (nezdolik) [inline source/extensions/config_subscription/grpc/grpc_mux_impl.cc:312]: would this result in error status if ratelimit config is not set?

File: source/extensions/config_subscription/grpc/new_grpc_mux_impl.cc
Comment (nezdolik) [inline source/extensions/config_subscription/grpc/new_grpc_mux_impl.cc:287]: this method does not guarantee that stream establishment was successful? E.g. it attempts to establish new stream and returns `absl::OkStatus()` independently if attempt was successful or not.

File: source/extensions/config_subscription/grpc/xds_mux/grpc_mux_impl.h
Comment (nezdolik) [inline source/extensions/config_subscription/grpc/xds_mux/grpc_mux_impl.h:266]: could this be factored out as constant and method will return string_view instead?

File: source/extensions/config_subscription/grpc/grpc_mux_failover.h
Comment (adisuissa) [reply source/extensions/config_subscription/grpc/grpc_mux_failover.h]: No it is not. Removed.
Thanks for the great comments!

File: source/extensions/config_subscription/grpc/grpc_mux_impl.cc
Comment (adisuissa) [reply source/extensions/config_subscription/grpc/grpc_mux_impl.cc]: I think the Envoy style guide allows both, but I understand the motivation behind this.
Updated.

File: source/extensions/config_subscription/grpc/new_grpc_mux_impl.cc
Comment (adisuissa) [reply source/extensions/config_subscription/grpc/new_grpc_mux_impl.cc:287]: No, the method does not guarantee that the stream establishment was successful.
The model of the gRPC-Muxes is as follows:
The gRPC-Mux receives a grpc-async-client, creates a grpc-stream (or grpc-failover) object and delegates the connection (stream establishment) and reconnection to that object.

File: source/extensions/config_subscription/grpc/xds_mux/grpc_mux_impl.h
Comment (adisuissa) [reply source/extensions/config_subscription/grpc/xds_mux/grpc_mux_impl.h:266]: returning an absl::string_view.
Regarding the constant, I have a vague recollection from my early C days, that the compiler puts the fixed strings from the code in the static data section of the process. It should then point to that contents in memory in the rest of the code (this was true for `const char*`, and I assume it will be the same for string_view).

File: source/extensions/config_subscription/grpc/grpc_mux_impl.cc
Comment (nezdolik) [reply source/extensions/config_subscription/grpc/grpc_mux_impl.cc:308]: and this is because ratelimit config is required? could there be use case where user does not need ratelimit config?

**Review by nezdolik — APPROVED** [review]
lgtm module one question about `rate_limit_settings`

File: source/extensions/config_subscription/grpc/grpc_mux_impl.cc
Comment (adisuissa) [reply source/extensions/config_subscription/grpc/grpc_mux_impl.cc:308]: Rate-limiting is an optional field (see [rate_limit_settings field](https://github.com/envoyproxy/envoy/blob/a65a5b673e6ac0ef12a3d8c5c2cdd2be4295927b/api/envoy/config/core/v3/config_source.proto#L107)).
There are default values that are returned if the rate-limit is empty (see [RateLimitSettings struct](https://github.com/envoyproxy/envoy/blob/a65a5b673e6ac0ef12a3d8c5c2cdd2be4295927b/source/common/config/utility.h#L51) and [parseRateLimitSettings() method](https://github.com/envoyproxy/envoy/blob/a65a5b673e6ac0ef12a3d8c5c2cdd2be4295927b/source/common/config/utility.cc#L206)).

Comment (adisuissa) [reply source/extensions/config_subscription/grpc/grpc_mux_impl.cc:308]: I guess my answer should be: it is not an error to have an empty rate-limit-settings field.

**Comment by jmarantz** [comment]
@wbpcode looks eady for senior stamp from you?

**Comment by jmarantz** [comment]
We have non-google mainter approval; over to Yan for senior stamp.

**Review by yanavlasov — APPROVED** [review]

**Comment by yanavlasov** [comment]
It looks like CI is stuck. But I have no way to unstuck it. Maybe merge main to see if ti fixes it?

**Pull request merged**
